collections/_posts/2022-10-03-MomentumResidualNeuralNetwork.md

```diff
@@ -0,0 +1,151 @@
+
+
+# Highlights of the paper
+
+* Exploration of Memory costs issues in residual neural network due to backpropagation schemes
+* Proposition of a new model, Momentum ResNets, which circumvents these memory issues by being invertible
+* Compared to other invertible models, this model is simple to integrate into the usual ResNets architectures and provides a rigorous mathematical setting.
+* Momentum ResNets can be interpreted as second order ordinary differential equations (ODEs).
+Momentum ResNets separate point clouds that ResNets fail to separate
+
+## Bibliography in the field 
+* Momentum in deep networks : (He et al., 2020; Chun et al., 2020; Nguyen et al., 2020; Li et al. 2018) 
+* Second-order models : (Norcliffe et al., 2020; Rusch & Mishra, 2021; Lu et al.,2018; Massaroli et al., 2020) 
+
+# Introduction
+
+The main goal of this paper is to explore the properties of a new model, Momentum ResNets, that circumvent the memory issues of ResNets by being invertible. It lies on the modification of the ResNet's forward rule.
+
+# Momentum Residual Neural Networks
+
+Backpropagation, used to optimize deep architectures, requires to store values at each layer during the network training.
+In classical ResNets, we have the feedforward relation:
+
+$$x_{n+1}=x_n+ f(x_n,\theta_n)$$
+
+where $$\theta_n$$ is the set of parameters, and $$x_n$$ the activations at layer $$n$$. 
+
+Memory issues occur when increasing the number of layers.
+
+The authors propose to use momentum equations that replace the classical relation above:
+
+$$ v_{n+1}=\gamma v_n+(1-\gamma)f(x_n,\theta_n)$$
+
+$$x_{n+1}=x_n+v_{n+1}$$
+
+where $$v$$ is a velocity term and $$\gamma$$ a momentum term.
+
+The method consists in modifying the forward equations using the same parameters as inputs. This is invertible since we can recover the values of $$x_{n}$$ and $$v_{n}$$ using $$x_{n+1}$$ and $$v_{n+1}$$.
+If we invert these equations, we get :
+
+$$x_n=x_{n+1}-v_{n+1}$$
+
+$$v_n=\frac{1}{\gamma} (v_{n+1}-(1-\gamma)f(x_n,\theta_n))$$
+
+This avoids the memory issues occuring due to the backpropagation step.
+
+Note: Momentum gradient descent is an alternative to classical gradient descent algorithms using second order partial derivatives.
+A overview of gradient descent optimization algorithms is given in the paper of S.Ruder (http://arxiv.org/abs/1609.04747) as mentioned in the paper.
+
+## Memory cost
+For usual ResNets, one need to store the weights of the network and the values of all activations for the training set at each layer. The memory needed is $$O(k*d*n_{batch})$$ while for the Momentum ResNets the memory need is $$O((1-\gamma)*k*d*n_{batch})$$.
+
+![](/collections/images/MomentumResNets/fig_memory_momentum.jpg)
+
+## The role of momentum
+When $$\gamma=0$$, they get a Classical ResNet. When $$\gamma \rightarrow 1$$, they get a special case of the invertible RevNet (Gomez et al, 2017). 
```

Comment:
When gamma = 0, the equation of v_n is wrong since we have the factor 1/gamma ;)